### PR TITLE
JS: add query js/incomplete-url-substring-sanitization

### PR DESCRIPTION
**Pull request by esben-semmle**
This adds a query for flagging insecure substring checks for the host of an URL.

The tricky part of the query is to recognize such substring checks precisely. I have experimented with identifying likely URL and hostname values, based on associated variable and property names, but it turns out that it is sufficently precise to recognize strings as likely URLs or host names with a regular expression. We may want to downgrade the precision or sharpen further once we see the results for all of LGTM.com.

Notice that this query excludes non-trivial position checks on result of `indexOf` calls, so it should not overlap with `js/incorrect-suffix-check`.

[Performance and results](https://git.semmle.com/esben/dist-compare-reports/tree/js/incomplete-url-sanitization_1544011269537) are reasonable.

### CODE REVIEW

File: javascript/ql/src/Security/CWE-020/IncompleteUrlSubstringSanitization.ql [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+          any(ArithmeticExpr e).getAnOperand().getUnderlyingValue() = call.asExpr()
+          or
+          // non-trivial position check on the indexOf-result
+          exists(EqualityTest test, Expr n |
+            test.hasOperands(call.asExpr(), n) |
+            not n.getIntValue() = [-1..0]
+          )
+        )
+        or
+        // the leading dot in for subdomain name sequence makes the suffix-check safe (if it is performed on the host of the url)
```
Comment (asger-semmle) [inline javascript/ql/src/Security/CWE-020/IncompleteUrlSubstringSanitization.ql]: > ... in for subdomain

typo?

**Comment by esben-semmle** [comment]
>  HTTP protocol should be a strong enough indicator on its own

I agree, but there are many benign `u.indexOf('https://')` checks, so I also require a TLD to be present -- that complements the fixed set of TLDs nicely.

**Comment by esben-semmle** [comment]
Ping @mc-semmle  for a doc review.

**Comment by esben-semmle** [comment]
Thanks. The changes are here: https://github.com/Semmle/ql/compare/10e5c063bcb30e49efd504e7b4c30bf86d5b98fd..4f5341139751c48b701f7da74ba8df26ca49229c

**Comment by esben-semmle** [comment]
Thanks. The changes are here: https://github.com/Semmle/ql/compare/10e5c063bcb30e49efd504e7b4c30bf86d5b98fd..4f5341139751c48b701f7da74ba8df26ca49229c

**Comment by mc-semmle** [comment]
Thanks for your latest updates to the documentation @esben-semmle - good to go from my point of view

**Pull request merged**